kafka-impl/src/main/java/io/streamnative/pulsar/handlers/kop/KafkaCommandDecoder.java

```diff
@@ -171,18 +177,29 @@ public void channelRead(ChannelHandlerContext ctx, Object msg) throws Exception
 
             CompletableFuture<AbstractResponse> responseFuture = new CompletableFuture<>();
             responseFuture.whenComplete((response, e) -> {
-                // Do nothing if it's triggered when responseFuture is expired
+                if (e != null) {
+                    if (e instanceof CancellationException) {
+                        if (log.isDebugEnabled()) {
+                            log.debug("[{}] Request {} is cancelled",
+                                    ctx.channel(), kafkaHeaderAndRequest.getHeader());
+                        }
+                    } else {
+                        log.error("[{}] Request {} is completed with exception",
+                                ctx.channel(), kafkaHeaderAndRequest.getHeader(), e);
+                    }
+                    return;
+                }
                 if (response == null) {
+                    log.error("[{}] Unexpected null completed future for request {}",
+                            ctx.channel(), kafkaHeaderAndRequest.getHeader());
```

Comment:
when response is null or cancelled by timeout, whether we should return an exception to client, otherwise, the client doesn't know what happens in the server side, and no error code returned to client.

Reply:
There're two cases that the request is cancelled:
1. Channel is closing. We needn't to send a response because Kafka client will know the channel is closed after server side calls `ctx.close()`.
2. The request is expired. We added a server side `request.timeout.ms` before, so it's triggered in `writeAndFlushResponseToClient` immediately. But I still didn't send back any response when it's expired. Because it's just to avoid blocking the followed response futures. Client side has its own `request.timeout.ms` that should be the **real** timeout.

As for when response is null, actually there's no possibility that response will be null from the code analysis. I check it only for the unexpected case and still rely on the client's own `request.timeout.ms`.

Reply:
What will happen when client configured `request.timeout.ms` value greater than server side?